spatial: st_segmentize goes out of bounds on Z/M coords

[(roachtest).sqlsmith/setup=tpcc/setting=default failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2795641&tab=buildLog) on [master@3d19b2cf6b290a152b23722fc32e995eed3b437b](https://github.com/cockroachdb/cockroach/commits/3d19b2cf6b290a152b23722fc32e995eed3b437b):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=default/run_1
	sqlsmith.go:218,sqlsmith.go:253,test_runner.go:768: error: pq: internal error: runtime error: index out of range [15] with length 15
		stmt:
		WITH
			with_29 (col_243)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(e'\x04*\x04':::STRING),
								(e'Zk\x0f"\r':::STRING),
								((NULL::TIMESTAMP || '':::STRING::STRING)::STRING),
								('|e':::STRING)
						)
							AS tab_125 (col_243)
				),
			with_30 (col_244) AS (SELECT * FROM (VALUES (2406307256:::OID), (NULL)) AS tab_126 (col_244))
		SELECT
			count(*) AS col_245
		FROM
			defaultdb.public.order_line@order_line_stock_fk_idx AS tab_127
		WHERE
			NULL
		GROUP BY
			tab_127.ol_supply_w_id
		HAVING
			st_intersects('010100008008F544DEA68002C24848B7372B3EFAC1B4D701DBF187EC41':::GEOMETRY::GEOMETRY, st_segmentize('010500004002000000010200004003000000A8C29CD554F3D2418853FF97E5C5FAC10B8156ABEDA7F3C1E8683C1DE252EF413739717AD77100C21C2AB284E605F44100E8746E29548C410099F27AF096E4418C98902C8018F8C1010200004002000000B267F30BEC7CF3C1601A79C86681E6415CF936003112FB41F3532DB9B9DE01C2F092845CF7BDD541BA662D74E55BF541':::GEOMETRY::GEOMETRY, 1.401298464324817e-45:::FLOAT8::FLOAT8)::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			58:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpcc/setting=default](https://teamcity.cockroachdb.com/viewLog.html?buildId=2795641&tab=artifacts#/sqlsmith/setup=tpcc/setting=default)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Ddefault.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/geo/geogfn/segmentize.go
+++ b/pkg/geo/geogfn/segmentize.go
@@ -57,6 +57,13 @@ func Segmentize(geography geo.Geography, segmentMaxLength float64) (geo.Geograph
 // segments is the power of 2, and all the segments are of the same length.
 // Note: List of points does not consist of end point.
 func segmentizeCoords(a geom.Coord, b geom.Coord, segmentMaxAngle float64) ([]float64, error) {
+	if len(a) != len(b) {
+		return nil, errors.Newf("cannot segmentize two coordinates of different dimensions")
+	}
+	if segmentMaxAngle <= 0 {
+		return nil, errors.Newf("maximum segment angle must be positive")
+	}
+
 	// Converted geom.Coord into s2.Point so we can segmentize the coordinates.
 	pointA := s2.PointFromLatLng(s2.LatLngFromDegrees(a.Y(), a.X()))
 	pointB := s2.PointFromLatLng(s2.LatLngFromDegrees(b.Y(), b.X()))
@@ -65,19 +72,17 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, segmentMaxAngle float64) ([]fl
 	// PostGIS' behavior appears to involve cutting this down into segments divisible
 	// by a power of two. As such, we do not use ceil(chordAngleBetweenPoints/segmentMaxAngle).
 	//
-	// This calculation is to determine the total number of segment between given
-	// 2 coordinates, ensuring that the segments are divided into parts divisible by
-	// a power of 2.
+	// This calculation determines the smallest power of 2 that is greater
+	// than ceil(chordAngleBetweenPoints/segmentMaxAngle).
 	//
-	// For that fraction by segment must be less than or equal to
-	// the fraction of max segment length to distance between point, since the
-	// total number of segment must be power of 2 therefore we can write as
-	// 1 / (2^n)[numberOfSegmentsToCreate] <= segmentMaxLength / distanceBetweenPoints < 1 / (2^(n-1))
-	// (2^n)[numberOfSegmentsToCreate] >= distanceBetweenPoints / segmentMaxLength > 2^(n-1)
-	// therefore n = ceil(log2(segmentMaxLength/distanceBetweenPoints)). Hence
-	// numberOfSegmentsToCreate = 2^(ceil(log2(segmentMaxLength/distanceBetweenPoints))).
+	// We can write that power as 2^n in the following inequality
+	// 2^n >= ceil(chordAngleBetweenPoints/segmentMaxLength) > 2^(n-1).
+	// We can drop the ceil since 2^n must be an int
+	// 2^n >= chordAngleBetweenPoints/segmentMaxLength > 2^(n-1).
+	// Then n = ceil(log2(chordAngleBetweenPoints/segmentMaxLength)).
+	// Hence numberOfSegmentsToCreate = 2^(ceil(log2(chordAngleBetweenPoints/segmentMaxLength))).
 	numberOfSegmentsToCreate := int(math.Pow(2, math.Ceil(math.Log2(chordAngleBetweenPoints/segmentMaxAngle))))
-	numPoints := 2 * (1 + numberOfSegmentsToCreate)
+	numPoints := len(a) * (1 + numberOfSegmentsToCreate)
 	if numPoints > geo.MaxAllowedSplitPoints {
 		return nil, errors.Newf(
 			"attempting to segmentize into too many coordinates; need %d points between %v and %v, max %d",
@@ -88,11 +93,19 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, segmentMaxAngle float64) ([]fl
 		)
 	}
 	allSegmentizedCoordinates := make([]float64, 0, numPoints)
-	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.X(), a.Y())
+	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.Clone()...)
+	segmentFraction := 1.0 / float64(numberOfSegmentsToCreate)
 	for pointInserted := 1; pointInserted < numberOfSegmentsToCreate; pointInserted++ {
 		newPoint := s2.Interpolate(float64(pointInserted)/float64(numberOfSegmentsToCreate), pointA, pointB)
 		latLng := s2.LatLngFromPoint(newPoint)
 		allSegmentizedCoordinates = append(allSegmentizedCoordinates, latLng.Lng.Degrees(), latLng.Lat.Degrees())
+		// Linearly interpolate Z and/or M coordinates.
+		for i := 2; i < len(a); i++ {
+			allSegmentizedCoordinates = append(
+				allSegmentizedCoordinates,
+				a[i]*(1-float64(pointInserted)*segmentFraction)+b[i]*(float64(pointInserted)*segmentFraction),
+			)
+		}
 	}
 	return allSegmentizedCoordinates, nil
 }

--- a/pkg/geo/geogfn/segmentize_test.go
+++ b/pkg/geo/geogfn/segmentize_test.go
@@ -95,6 +95,11 @@ func TestSegmentize(t *testing.T) {
 			maxSegmentLength: -1,
 			expectedWKT:      "MULTIPOINT((0 0), (1 1))",
 		},
+		{
+			wkt:              "LINESTRING(0 0 25, 0 1 0, 2 5 100)",
+			maxSegmentLength: 150000.0,
+			expectedWKT:      "LINESTRING Z (0 0 25,0 1 0,0.49878052093921765 2.0003038990352664 25,0.9981696941692514 3.0004561476391296 50,1.4984735304805308 4.000380457593079 75,2 5 100)",
+		},
 	}
 	for _, test := range segmentizeTestCases {
 		t.Run(fmt.Sprintf("%s, maximum segment length: %f", test.wkt, test.maxSegmentLength), func(t *testing.T) {
@@ -131,57 +136,93 @@ func TestSegmentizeCoords(t *testing.T) {
 		desc                 string
 		a                    geom.Coord
 		b                    geom.Coord
-		segmentMaxLength     float64
+		segmentMaxAngle      float64
 		resultantCoordinates []float64
 	}{
 		{
 			desc:                 `Coordinate(0, 0) to Coordinate(85, 85), 0.781600222084459`,
 			a:                    geom.Coord{0, 0},
 			b:                    geom.Coord{85, 85},
-			segmentMaxLength:     0.781600222084459,
+			segmentMaxAngle:      0.781600222084459,
 			resultantCoordinates: []float64{0, 0, 4.924985039227315, 44.568038920632105},
 		},
 		{
 			desc:                 `Coordinate(0, 0) to Coordinate(85, 85), 0.7816000651234446`,
 			a:                    geom.Coord{0, 0},
 			b:                    geom.Coord{85, 85},
-			segmentMaxLength:     0.7816000651234446,
+			segmentMaxAngle:      0.7816000651234446,
 			resultantCoordinates: []float64{0, 0, 2.0486953806277866, 22.302074138733936, 4.924985039227315, 44.568038920632105, 11.655816669136822, 66.66485041602017},
 		},
 		{
 			desc:                 `Coordinate(85, 85) to Coordinate(0, 0), 0.29502092024628396`,
 			a:                    geom.Coord{85, 85},
 			b:                    geom.Coord{0, 0},
-			segmentMaxLength:     0.29502092024628396,
+			segmentMaxAngle:      0.29502092024628396,
 			resultantCoordinates: []float64{85, 85, 22.871662720021178, 77.3609628116894, 11.655816669136824, 66.66485041602017, 7.329091976767396, 55.658764687902504, 4.924985039227315, 44.56803892063211, 3.299940624127866, 33.443216802941045, 2.048695380627787, 22.302074138733943, 0.9845421446758968, 11.1527721155093},
-		},
-		{
-			desc:                 `Coordinate(85, 85) to Coordinate(0, 0), -1`,
-			a:                    geom.Coord{85, 85},
-			b:                    geom.Coord{0, 0},
-			segmentMaxLength:     -1,
-			resultantCoordinates: []float64{85, 85},
 		},
 		{
 			desc:                 `Coordinate(0, 0) to Coordinate(0, 0), 0.29`,
 			a:                    geom.Coord{0, 0},
 			b:                    geom.Coord{0, 0},
-			segmentMaxLength:     0.29,
+			segmentMaxAngle:      0.29,
 			resultantCoordinates: []float64{0, 0},
 		},
 		{
 			desc:                 `Coordinate(85, 85) to Coordinate(0, 0), 1.563200444168918`,
 			a:                    geom.Coord{85, 85},
 			b:                    geom.Coord{0, 0},
-			segmentMaxLength:     1.563200444168918,
+			segmentMaxAngle:      1.563200444168918,
 			resultantCoordinates: []float64{85, 85},
+		},
+		{
+			desc:                 `Coordinate(0, 16, 23, 10) to Coordinate(1, 0, -5, 0), 0.07848050723825097`,
+			a:                    geom.Coord{0, 16, 23, 10},
+			b:                    geom.Coord{1, 0, -5, 0},
+			segmentMaxAngle:      0.07848050723825097,
+			resultantCoordinates: []float64{0, 16, 23, 10, 0.2587272349066411, 12.000265604969918, 16, 7.5, 0.5098761137159087, 8.00030056571944, 9, 5, 0.7561364483510505, 4.000186752673144, 2, 2.5},
 		},
 	}
 	for _, test := range testCases {
 		t.Run(test.desc, func(t *testing.T) {
-			convertedPoints, err := segmentizeCoords(test.a, test.b, test.segmentMaxLength)
+			convertedPoints, err := segmentizeCoords(test.a, test.b, test.segmentMaxAngle)
 			require.NoError(t, err)
 			require.Equal(t, test.resultantCoordinates, convertedPoints)
+		})
+	}
+
+	errorTestCases := []struct {
+		desc            string
+		a               geom.Coord
+		b               geom.Coord
+		segmentMaxAngle float64
+		expectedErr     string
+	}{
+		{
+			desc:            "too many segments required",
+			a:               geom.Coord{0, 100},
+			b:               geom.Coord{1, 1},
+			segmentMaxAngle: 2e-8,
+			expectedErr:     fmt.Sprintf("attempting to segmentize into too many coordinates; need 268435458 points between [0 100] and [1 1], max %d", geo.MaxAllowedSplitPoints),
+		},
+		{
+			desc:            "input coords have different dimensions",
+			a:               geom.Coord{1, 1},
+			b:               geom.Coord{1, 2, 3},
+			segmentMaxAngle: 1,
+			expectedErr:     "cannot segmentize two coordinates of different dimensions",
+		},
+		{
+			desc:            "negative max segment angle",
+			a:               geom.Coord{1, 1},
+			b:               geom.Coord{2, 2},
+			segmentMaxAngle: -1,
+			expectedErr:     "maximum segment angle must be positive",
+		},
+	}
+	for _, test := range errorTestCases {
+		t.Run(test.desc, func(t *testing.T) {
+			_, err := segmentizeCoords(test.a, test.b, test.segmentMaxAngle)
+			require.EqualError(t, err, test.expectedErr)
 		})
 	}
 }

--- a/pkg/geo/geomfn/segmentize.go
+++ b/pkg/geo/geomfn/segmentize.go
@@ -51,12 +51,20 @@ func Segmentize(g geo.Geometry, segmentMaxLength float64) (geo.Geometry, error) 
 // segment has a length less than or equal to given maximum segment length.
 // Note: List of points does not consist of end point.
 func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) ([]float64, error) {
+	if len(a) != len(b) {
+		return nil, errors.Newf("cannot segmentize two coordinates of different dimensions")
+	}
+	if maxSegmentLength <= 0 {
+		return nil, errors.Newf("maximum segment length must be positive")
+	}
+
+	// Only 2D distance is considered for determining number of segments.
 	distanceBetweenPoints := math.Sqrt(math.Pow(a.X()-b.X(), 2) + math.Pow(b.Y()-a.Y(), 2))
 
 	// numberOfSegmentsToCreate represent the total number of segments
 	// in which given two coordinates will be divided.
 	numberOfSegmentsToCreate := int(math.Ceil(distanceBetweenPoints / maxSegmentLength))
-	numPoints := 2 * (1 + numberOfSegmentsToCreate)
+	numPoints := len(a) * (1 + numberOfSegmentsToCreate)
 	if numPoints > geo.MaxAllowedSplitPoints {
 		return nil, errors.Newf(
 			"attempting to segmentize into too many coordinates; need %d points between %v and %v, max %d",
@@ -67,14 +75,20 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) ([]f
 		)
 	} // segmentFraction represent the fraction of length each segment
 	// has with respect to total length between two coordinates.
-	allSegmentizedCoordinates := make([]float64, 0, 2*(1+numberOfSegmentsToCreate))
+	allSegmentizedCoordinates := make([]float64, 0, numPoints)
 	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.Clone()...)
 	segmentFraction := 1.0 / float64(numberOfSegmentsToCreate)
 	for pointInserted := 1; pointInserted < numberOfSegmentsToCreate; pointInserted++ {
+		segmentPoint := make([]float64, 0, len(a))
+		for i := 0; i < len(a); i++ {
+			segmentPoint = append(
+				segmentPoint,
+				a[i]*(1-float64(pointInserted)*segmentFraction)+b[i]*(float64(pointInserted)*segmentFraction),
+			)
+		}
 		allSegmentizedCoordinates = append(
 			allSegmentizedCoordinates,
-			b.X()*float64(pointInserted)*segmentFraction+a.X()*(1-float64(pointInserted)*segmentFraction),
-			b.Y()*float64(pointInserted)*segmentFraction+a.Y()*(1-float64(pointInserted)*segmentFraction),
+			segmentPoint...,
 		)
 	}
 

--- a/pkg/geo/geomfn/segmentize_test.go
+++ b/pkg/geo/geomfn/segmentize_test.go
@@ -38,7 +38,7 @@ func TestSegmentize(t *testing.T) {
 		{
 			wkt:              "LINESTRING (1.0 1.0, 2.0 2.0, 3.0 3.0)",
 			maxSegmentLength: 0.33333,
-			expectedWKT:      "LINESTRING (1.0 1.0, 1.2000000000000002 1.2000000000000002, 1.4 1.4, 1.6 1.6, 1.8 1.8, 2.0 2.0, 2.2 2.2, 2.4000000000000004 2.4000000000000004, 2.5999999999999996 2.5999999999999996, 2.8000000000000003 2.8000000000000003, 3 3)",
+			expectedWKT:      "LINESTRING (1.0 1.0, 1.2000000000000002 1.2000000000000002, 1.4 1.4, 1.6 1.6, 1.8 1.8, 2.0 2.0, 2.2 2.2, 2.4000000000000004 2.4000000000000004, 2.6 2.6, 2.8000000000000003 2.8000000000000003, 3 3)",
 		},
 		{
 			wkt:              "LINESTRING EMPTY",
@@ -54,6 +54,16 @@ func TestSegmentize(t *testing.T) {
 			wkt:              "LINESTRING (0.0 0.0, 0.0 10.0, 0.0 16.0)",
 			maxSegmentLength: 3,
 			expectedWKT:      "LINESTRING (0.0 0.0,0.0 2.5,0.0 5.0,0.0 7.5,0.0 10.0,0.0 13.0,0.0 16.0)",
+		},
+		{
+			wkt:              "LINESTRING M (0 0 23, 1 0 -5)",
+			maxSegmentLength: 0.25,
+			expectedWKT:      "LINESTRING M (0 0 23, 0.25 0 16, 0.5 0 9, 0.75 0 2, 1 0 -5)",
+		},
+		{
+			wkt:              "LINESTRING ZM (0 0 23 10, 1 0 -5 0)",
+			maxSegmentLength: 0.5,
+			expectedWKT:      "LINESTRING ZM (0 0 23 10,0.5 0 9 5,1 0 -5 0)",
 		},
 		{
 			wkt:              "POLYGON ((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 0.0))",
@@ -74,6 +84,16 @@ func TestSegmentize(t *testing.T) {
 			wkt:              "POLYGON EMPTY",
 			maxSegmentLength: 1,
 			expectedWKT:      "POLYGON EMPTY",
+		},
+		{
+			wkt:              "POLYGON Z EMPTY",
+			maxSegmentLength: 1,
+			expectedWKT:      "POLYGON Z EMPTY",
+		},
+		{
+			wkt:              "POLYGON Z ((1 1 0, 5 1 10, 5 3 20, 1 3 30, 1 1 0))",
+			maxSegmentLength: 2,
+			expectedWKT:      "POLYGON Z ((1 1 0, 3 1 5, 5 1 10, 5 3 20, 3 3 25, 1 3 30, 1 1 0))",
 		},
 		{
 			wkt:              "MULTIPOINT ((1.0 1.0), (2.0 2.0))",
@@ -156,13 +176,6 @@ func TestSegmentizeCoords(t *testing.T) {
 			resultantCoordinates: []float64{0, 0, 0.3333333333333333, 0, 0.6666666666666666, 0},
 		},
 		{
-			desc:                 `Coordinate(1, 1) to Coordinate(0, 0), -1`,
-			a:                    geom.Coord{1, 1},
-			b:                    geom.Coord{0, 0},
-			segmentMaxLength:     -1,
-			resultantCoordinates: []float64{1, 1},
-		},
-		{
 			desc:                 `Coordinate(1, 1) to Coordinate(0, 0), 2`,
 			a:                    geom.Coord{1, 1},
 			b:                    geom.Coord{0, 0},
@@ -185,13 +198,39 @@ func TestSegmentizeCoords(t *testing.T) {
 		})
 	}
 
-	t.Run("many coordinates to segmentize", func(t *testing.T) {
-		g := geo.MustParseGeometry("LINESTRING(0 0, 100 100)")
-		_, err := Segmentize(g, 0.001)
-		require.EqualError(
-			t,
-			err,
-			fmt.Sprintf("attempting to segmentize into too many coordinates; need 282846 points between [0 0] and [100 100], max %d", geo.MaxAllowedSplitPoints),
-		)
-	})
+	errorTestCases := []struct {
+		desc             string
+		a                geom.Coord
+		b                geom.Coord
+		segmentMaxLength float64
+		expectedErr      string
+	}{
+		{
+			desc:             "too many segments required",
+			a:                geom.Coord{0, 0},
+			b:                geom.Coord{100, 100},
+			segmentMaxLength: 0.001,
+			expectedErr:      fmt.Sprintf("attempting to segmentize into too many coordinates; need 282846 points between [0 0] and [100 100], max %d", geo.MaxAllowedSplitPoints),
+		},
+		{
+			desc:             "input coords have different dimensions",
+			a:                geom.Coord{1, 1},
+			b:                geom.Coord{1, 2, 3},
+			segmentMaxLength: 1,
+			expectedErr:      "cannot segmentize two coordinates of different dimensions",
+		},
+		{
+			desc:             "negative max segment length",
+			a:                geom.Coord{1, 1},
+			b:                geom.Coord{2, 2},
+			segmentMaxLength: -1,
+			expectedErr:      "maximum segment length must be positive",
+		},
+	}
+	for _, test := range errorTestCases {
+		t.Run(test.desc, func(t *testing.T) {
+			_, err := segmentizeCoords(test.a, test.b, test.segmentMaxLength)
+			require.EqualError(t, err, test.expectedErr)
+		})
+	}
 }

--- a/pkg/geo/geosegmentize/geosegmentize.go
+++ b/pkg/geo/geosegmentize/geosegmentize.go
@@ -33,6 +33,7 @@ func Segmentize(
 	if geometry.Empty() {
 		return geometry, nil
 	}
+	layout := geometry.Layout()
 	switch geometry := geometry.(type) {
 	case *geom.Point, *geom.MultiPoint:
 		return geometry, nil
@@ -50,9 +51,9 @@ func Segmentize(
 		}
 		// Appending end point as it wasn't included in the iteration of coordinates.
 		allFlatCoordinates = append(allFlatCoordinates, geometry.Coord(geometry.NumCoords()-1)...)
-		return geom.NewLineStringFlat(geom.XY, allFlatCoordinates).SetSRID(geometry.SRID()), nil
+		return geom.NewLineStringFlat(layout, allFlatCoordinates).SetSRID(geometry.SRID()), nil
 	case *geom.MultiLineString:
-		segMultiLine := geom.NewMultiLineString(geom.XY).SetSRID(geometry.SRID())
+		segMultiLine := geom.NewMultiLineString(layout).SetSRID(geometry.SRID())
 		for lineIdx := 0; lineIdx < geometry.NumLineStrings(); lineIdx++ {
 			l, err := Segmentize(geometry.LineString(lineIdx), segmentMaxAngleOrLength, segmentizeCoords)
 			if err != nil {
@@ -78,9 +79,9 @@ func Segmentize(
 		}
 		// Appending end point as it wasn't included in the iteration of coordinates.
 		allFlatCoordinates = append(allFlatCoordinates, geometry.Coord(geometry.NumCoords()-1)...)
-		return geom.NewLinearRingFlat(geom.XY, allFlatCoordinates).SetSRID(geometry.SRID()), nil
+		return geom.NewLinearRingFlat(layout, allFlatCoordinates).SetSRID(geometry.SRID()), nil
 	case *geom.Polygon:
-		segPolygon := geom.NewPolygon(geom.XY).SetSRID(geometry.SRID())
+		segPolygon := geom.NewPolygon(layout).SetSRID(geometry.SRID())
 		for loopIdx := 0; loopIdx < geometry.NumLinearRings(); loopIdx++ {
 			l, err := Segmentize(geometry.LinearRing(loopIdx), segmentMaxAngleOrLength, segmentizeCoords)
 			if err != nil {
@@ -93,7 +94,7 @@ func Segmentize(
 		}
 		return segPolygon, nil
 	case *geom.MultiPolygon:
-		segMultiPolygon := geom.NewMultiPolygon(geom.XY).SetSRID(geometry.SRID())
+		segMultiPolygon := geom.NewMultiPolygon(layout).SetSRID(geometry.SRID())
 		for polygonIdx := 0; polygonIdx < geometry.NumPolygons(); polygonIdx++ {
 			p, err := Segmentize(geometry.Polygon(polygonIdx), segmentMaxAngleOrLength, segmentizeCoords)
 			if err != nil {
@@ -107,6 +108,10 @@ func Segmentize(
 		return segMultiPolygon, nil
 	case *geom.GeometryCollection:
 		segGeomCollection := geom.NewGeometryCollection().SetSRID(geometry.SRID())
+		err := segGeomCollection.SetLayout(layout)
+		if err != nil {
+			return nil, err
+		}
 		for geoIdx := 0; geoIdx < geometry.NumGeoms(); geoIdx++ {
 			g, err := Segmentize(geometry.Geom(geoIdx), segmentMaxAngleOrLength, segmentizeCoords)
 			if err != nil {

